File: nodejs/test/e2e/permissions.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { readFile, writeFile } from "fs/promises";
import { join } from "path";
import { describe, expect, it } from "vitest";
import type { PermissionRequest, PermissionRequestResult } from "../../src/index.js";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Permission callbacks", async () => {
    const { copilotClient: client, workDir } = await createSdkTestContext();

    it("should invoke permission handler for write operations", async () => {
        const permissionRequests: PermissionRequest[] = [];

        const session = await client.createSession({
            onPermissionRequest: (request, invocation) => {
                permissionRequests.push(request);
                expect(invocation.sessionId).toBe(session.sessionId);

                // Approve the permission
                const result: PermissionRequestResult = { kind: "approved" };
                return result;
            },
        });

        await writeFile(join(workDir, "test.txt"), "original content");

        await session.sendAndWait({
            prompt: "Edit test.txt and replace 'original' with 'modified'",
        });

        // Should have received at least one permission request
        expect(permissionRequests.length).toBeGreaterThan(0);

        // Should include write permission request
        const writeRequests = permissionRequests.filter((req) => req.kind === "write");
        expect(writeRequests.length).toBeGreaterThan(0);

        await session.destroy();
    });

    it("should deny permission when handler returns denied", async () => {
        const session = await client.createSession({
            onPermissionRequest: () => {
                return { kind: "denied-interactively-by-user" };
            },
        });

        const originalContent = "protected content";
        const testFile = join(workDir, "protected.txt");
        await writeFile(testFile, originalContent);

        await session.sendAndWait({
            prompt: "Edit protected.txt and replace 'protected' with 'hacked'.",
        });

        // Verify the file was NOT modified
        const content = await readFile(testFile, "utf-8");
        expect(content).toBe(originalContent);

        await session.destroy();
    });

    it("should deny tool operations by default when no handler is provided", async () => {
        let permissionDenied = false;

        const session = await client.createSession();
        session.on((event) => {
            if (
                event.type === "tool.execution_complete" &&
                !event.data.success &&
                event.data.error?.message.includes("Permission denied")
            ) {
                permissionDenied = true;
            }
        });

        await session.sendAndWait({ prompt: "Run 'node --version'" });

        expect(permissionDenied).toBe(true);

        await session.destroy();
    });

    it("should deny tool operations by default when no handler is provided after resume", async () => {
        const session1 = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session1.sessionId;
        await session1.sendAndWait({ prompt: "What is 1+1?" });

        const session2 = await client.resumeSession(sessionId);
        let permissionDenied = false;
        session2.on((event) => {
            if (
                event.type === "tool.execution_complete" &&
                !event.data.success &&
                event.data.error?.message.includes("Permission denied")
            ) {
                permissionDenied = true;
            }
        });

        await session2.sendAndWait({ prompt: "Run 'node --version'" });

        expect(permissionDenied).toBe(true);

        await session2.destroy();
    });

    it("should work without permission handler (default behavior)", async () => {
        // Create session without onPermissionRequest handler
        const session = await client.createSession();

        const message = await session.sendAndWait({
            prompt: "What is 2+2?",
        });
        expect(message?.data.content).toContain("4");

        await session.destroy();
    });

    it("should handle async permission handler", async () => {
        const permissionRequests: PermissionRequest[] = [];

        const session = await client.createSession({
            onPermissionRequest: async (request, _invocation) => {
                permissionRequests.push(request);

                // Simulate async permission check (e.g., user prompt)
                await new Promise((resolve) => setTimeout(resolve, 10));

                return { kind: "approved" };
            },
        });

        await session.sendAndWait({
            prompt: "Run 'echo test' and tell me what happens",
        });

        expect(permissionRequests.length).toBeGreaterThan(0);

        await session.destroy();
    });

    it("should resume session with permission handler", async () => {
        const permissionRequests: PermissionRequest[] = [];

        // Create session without permission handler
        const session1 = await client.createSession();
        const sessionId = session1.sessionId;
        await session1.sendAndWait({ prompt: "What is 1+1?" });

        // Resume with permission handler
        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: (request) => {
                permissionRequests.push(request);
                return { kind: "approved" };
            },
        });

        await session2.sendAndWait({
            prompt: "Run 'echo resumed' for me",
        });

        // Should have permission requests from resumed session
        expect(permissionRequests.length).toBeGreaterThan(0);

        await session2.destroy();
    });

    it("should handle permission handler errors gracefully", async () => {
        const session = await client.createSession({
            onPermissionRequest: () => {
                throw new Error("Handler error");
            },
        });

        const message = await session.sendAndWait({
            prompt: "Run 'echo test'. If you can't, say 'failed'.",
        });

        // Should handle the error and deny permission
        expect(message?.data.content?.toLowerCase()).toMatch(/fail|cannot|unable|permission/);

        await session.destroy();
    });

    it("should receive toolCallId in permission requests", async () => {
        let receivedToolCallId = false;

        const session = await client.createSession({
            onPermissionRequest: (request) => {
                if (request.toolCallId) {
                    receivedToolCallId = true;
                    expect(typeof request.toolCallId).toBe("string");
                    expect(request.toolCallId.length).toBeGreaterThan(0);
                }
                return { kind: "approved" };
            },
        });

        await session.sendAndWait({
            prompt: "Run 'echo test'",
        });

        expect(receivedToolCallId).toBe(true);

        await session.destroy();
    });
});

File: java/src/test/java/com/github/copilot/PermissionsTest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static org.junit.jupiter.api.Assertions.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;

import com.github.copilot.generated.AssistantMessageEvent;
import com.github.copilot.generated.ToolExecutionCompleteEvent;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.PermissionRequest;
import com.github.copilot.rpc.PermissionRequestResult;
import com.github.copilot.rpc.PermissionRequestResultKind;
import com.github.copilot.rpc.SessionConfig;
import com.github.copilot.rpc.ResumeSessionConfig;
import com.github.copilot.rpc.MessageOptions;

/**
 * Tests for permission callback functionality.
 *
 * <p>
 * These tests use the shared CapiProxy infrastructure for deterministic API
 * response replay. Snapshots are stored in test/snapshots/permissions/.
 * </p>
 */
public class PermissionsTest {

    private static E2ETestContext ctx;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    /**
     * Verifies that permission handler is invoked for write operations.
     *
     * @see Snapshot: permissions/permission_handler_for_write_operations
     */
    @Test
    void testPermissionHandlerForWriteOperations(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "permission_handler_for_write_operations");

        var permissionRequests = new ArrayList<PermissionRequest>();

        final String[] sessionIdHolder = new String[1];

        var config = new SessionConfig().setOnPermissionRequest((request, invocation) -> {
            permissionRequests.add(request);
            assertEquals(sessionIdHolder[0], invocation.getSessionId());
            // Approve the permission
            return CompletableFuture
                    .completedFuture(new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED));
        });

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(config).get();
            sessionIdHolder[0] = session.getSessionId();

            // Write a test file
            Path testFile = ctx.getWorkDir().resolve("test.txt");
            Files.writeString(testFile, "original content");

            session.sendAndWait(new MessageOptions().setPrompt("Edit test.txt and replace 'original' with 'modified'"))
                    .get(60, TimeUnit.SECONDS);

            // Should have received at least one permission request
            assertFalse(permissionRequests.isEmpty(), "Should have received permission requests");

            // Should include write permission request
            boolean hasWriteRequest = permissionRequests.stream().anyMatch(req -> "write".equals(req.getKind()));
            assertTrue(hasWriteRequest, "Should have received a write permission request");

            session.close();
        }
    }

    /**
     * Verifies that permissions can be denied.
     *
     * @see Snapshot: permissions/deny_permission
     */
    @Test
    void testDenyPermission(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "deny_permission");

        var config = new SessionConfig().setOnPermissionRequest((request, invocation) -> {
            // Deny all permissions
            return CompletableFuture
                    .completedFuture(new PermissionRequestResult().setKind(PermissionRequestResultKind.REJECTED));
        });

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(config).get();

            String originalContent = "protected content";
            Path testFile = ctx.getWorkDir().resolve("protected.txt");
            Files.writeString(testFile, originalContent);

            session.sendAndWait(
                    new MessageOptions().setPrompt("Edit protected.txt and replace 'protected' with 'hacked'."))
                    .get(60, TimeUnit.SECONDS);

            // Verify the file was NOT modified
            String content = Files.readString(testFile);
            assertEquals(originalContent, content, "File should not have been modified");

            session.close();
        }
    }

    /**
     * Verifies that sessions work with the approve-all permission handler.
     *
     * @see Snapshot: permissions/should_work_with_approve_all_permission_handler
     */
    @Test
    void testShouldWorkWithApproveAllPermissionHandler(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "should_work_with_approve_all_permission_handler");

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();

            AssistantMessageEvent response = session.sendAndWait(new MessageOptions().setPrompt("What is 2+2?")).get(60,
                    TimeUnit.SECONDS);

            assertNotNull(response);
            assertTrue(response.getData().content().contains("4"),
                    "Response should contain 4: " + response.getData().content());

            session.close();
        }
    }

    /**
     * Verifies that async permission handlers work correctly.
     *
     * @see Snapshot: permissions/async_permission_handler
     */
    @Test
    void testAsyncPermissionHandler(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "async_permission_handler");

        var permissionRequests = new ArrayList<PermissionRequest>();

        var config = new SessionConfig().setOnPermissionRequest((request, invocation) -> {
            permissionRequests.add(request);

            // Simulate async permission check with delay
            return CompletableFuture.supplyAsync(() -> {
                try {
                    Thread.sleep(10); // Small delay to simulate async check
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
                return new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED);
            });
        });

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(config).get();

            session.sendAndWait(new MessageOptions().setPrompt("Run 'echo test' and tell me what happens")).get(60,
                    TimeUnit.SECONDS);

            // Should have received permission requests
            assertFalse(permissionRequests.isEmpty(), "Should have received permission requests");

            session.close();
        }
    }

    /**
     * Verifies that permission handlers work when resuming a session.
     *
     * @see Snapshot: permissions/resume_session_with_permission_handler
     */
    @Test
    void testResumeSessionWithPermissionHandler(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "resume_session_with_permission_handler");

        var permissionRequests = new ArrayList<PermissionRequest>();

        try (CopilotClient client = ctx.createClient()) {
            // Create session with approve-all handler for initial exchange
            CopilotSession session1 = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();
            String sessionId = session1.getSessionId();
            session1.sendAndWait(new MessageOptions().setPrompt("What is 1+1?")).get(60, TimeUnit.SECONDS);

            // Resume with permission handler
            var resumeConfig = new ResumeSessionConfig().setOnPermissionRequest((request, invocation) -> {
                permissionRequests.add(request);
                return CompletableFuture
                        .completedFuture(new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED));
            });

            CopilotSession session2 = client.resumeSession(sessionId, resumeConfig).get();

            assertEquals(sessionId, session2.getSessionId());

            session2.sendAndWait(new MessageOptions().setPrompt("Run 'echo resumed' for me")).get(60, TimeUnit.SECONDS);

            // Should have permission requests from resumed session
            assertFalse(permissionRequests.isEmpty(), "Should have received permission requests from resumed session");

            session2.close();
        }
    }

    /**
     * Verifies that tool call IDs are included in permission requests.
     *
     * @see Snapshot: permissions/tool_call_id_in_permission_requests
     */
    @Test
    void testToolCallIdInPermissionRequests(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "tool_call_id_in_permission_requests");

        final boolean[] receivedToolCallId = {false};

        var config = new SessionConfig().setOnPermissionRequest((request, invocation) -> {
            if (request.getToolCallId() != null) {
                receivedToolCallId[0] = true;
                assertFalse(request.getToolCallId().isEmpty(), "Tool call ID should not be empty");
            }
            return CompletableFuture
                    .completedFuture(new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED));
        });

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(config).get();

            session.sendAndWait(new MessageOptions().setPrompt("Run 'echo test'")).get(60, TimeUnit.SECONDS);

            assertTrue(receivedToolCallId[0], "Should have received toolCallId in permission request");

            session.close();
        }
    }

    /**
     * Verifies that permission handler errors are handled gracefully.
     * <p>
     * When the handler throws an exception, the SDK should deny the permission and
     * the assistant should indicate it couldn't complete the task.
     * </p>
     *
     * @see Snapshot: permissions/should_handle_permission_handler_errors_gracefully
     */
    @Test
    void testShouldHandlePermissionHandlerErrorsGracefully(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "should_handle_permission_handler_errors_gracefully");

        var config = new SessionConfig().setOnPermissionRequest((request, invocation) -> {
            // Throw an error in the handler
            throw new RuntimeException("Handler error");
        });

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(config).get();

            AssistantMessageEvent response = session
                    .sendAndWait(new MessageOptions().setPrompt("Run 'echo test'. If you can't, say 'failed'."))
                    .get(60, TimeUnit.SECONDS);

            // Should handle the error and deny permission
            assertNotNull(response);
            String content = response.getData().content().toLowerCase();
            assertTrue(content.contains("fail") || content.contains("cannot") || content.contains("unable")
                    || content.contains("permission"), "Response should indicate failure: " + content);

            session.close();
        }
    }

    /**
     * Verifies that tool operations are denied when the handler explicitly denies.
     *
     * @see Snapshot:
     *      permissions/should_deny_tool_operations_when_handler_explicitly_denies
     */
    @Test
    void testShouldDenyToolOperationsWhenHandlerExplicitlyDenies(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "should_deny_tool_operations_when_handler_explicitly_denies");

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client.createSession(new SessionConfig()
                    .setOnPermissionRequest((request, invocation) -> CompletableFuture.completedFuture(
                            new PermissionRequestResult().setKind(PermissionRequestResultKind.USER_NOT_AVAILABLE))))
                    .get();

            final boolean[] permissionDenied = {false};
            session.on(ToolExecutionCompleteEvent.class, evt -> {
                if (!evt.getData().success() && evt.getData().error() != null && evt.getData().error().message() != null
                        && evt.getData().error().message().contains("Permission denied")) {
                    permissionDenied[0] = true;
                }
            });

            session.sendAndWait(new MessageOptions().setPrompt("Run 'node --version'")).get(60, TimeUnit.SECONDS);

            assertTrue(permissionDenied[0], "Expected a tool.execution_complete event with Permission denied result");

            session.close();
        }
    }

    /**
     * Verifies that tool operations are denied when the handler explicitly denies
     * after resuming a session.
     *
     * @see Snapshot:
     *      permissions/should_deny_tool_operations_when_handler_explicitly_denies_after_resume
     */
    @Test
    void testShouldDenyToolOperationsWhenHandlerExplicitlyDeniesAfterResume(TestInfo testInfo) throws Exception {
        ctx.configureForTest("permissions", "should_deny_tool_operations_when_handler_explicitly_denies_after_resume");

        try (CopilotClient client = ctx.createClient()) {
            var config = new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL);
            CopilotSession session1 = client.createSession(config).get();
            String sessionId = session1.getSessionId();
            session1.sendAndWait(new MessageOptions().setPrompt("What is 1+1?")).get(60, TimeUnit.SECONDS);

            CopilotSession session2 = client.resumeSession(sessionId, new ResumeSessionConfig()
                    .setOnPermissionRequest((request, invocation) -> CompletableFuture.completedFuture(
                            new PermissionRequestResult().setKind(PermissionRequestResultKind.USER_NOT_AVAILABLE))))
                    .get();

            final boolean[] permissionDenied = {false};
            session2.on(ToolExecutionCompleteEvent.class, evt -> {
                if (!evt.getData().success() && evt.getData().error() != null && evt.getData().error().message() != null
                        && evt.getData().error().message().contains("Permission denied")) {
                    permissionDenied[0] = true;
                }
            });

            session2.sendAndWait(new MessageOptions().setPrompt("Run 'node --version'")).get(60, TimeUnit.SECONDS);

            assertTrue(permissionDenied[0], "Expected a tool.execution_complete event with Permission denied result");

            session2.close();
        }
    }

    /**
     * Verifies that a permission handler returning {@code noResult} is handled
     * correctly — the handler is called, and the session can be aborted afterward.
     *
     * @see Snapshot: permissions/should_deny_permission_with_noresult_kind
     */
    @Test
    void testShouldDenyPermissionWithNoResultKind() throws Exception {
        ctx.configureForTest("permissions", "should_deny_permission_with_noresult_kind");

        var permissionCalled = new CompletableFuture<Boolean>();

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest((request, invocation) -> {
                        permissionCalled.complete(true);
                        return CompletableFuture.completedFuture(
                                new PermissionRequestResult().setKind(PermissionRequestResultKind.NO_RESULT));
                    })).get();

            session.send(new MessageOptions().setPrompt("Run 'node --version'"));

            assertTrue(permissionCalled.get(30, TimeUnit.SECONDS),
                    "Expected the no-result permission handler to be called.");

            session.abort().get(10, TimeUnit.SECONDS);
            session.close();
        }
    }

    /**
     * Verifies that the runtime short-circuits the permission handler when
     * {@code session.permissions.setApproveAll(true)} has been called.
     *
     * @see Snapshot:
     *      permissions/should_short_circuit_permission_handler_when_set_approve_all_enabled
     */
    @Test
    void testShouldShortCircuitPermissionHandlerWhenSetApproveAllEnabled() throws Exception {
        ctx.configureForTest("permissions", "should_short_circuit_permission_handler_when_set_approve_all_enabled");

        var handlerCallCount = new int[]{0};

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest((request, invocation) -> {
                        handlerCallCount[0]++;
                        return CompletableFuture.completedFuture(
                                new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED));
                    })).get();

            // Set approve-all so the runtime short-circuits
            var setResult = session.getRpc().permissions
                    .setApproveAll(new com.github.copilot.generated.rpc.SessionPermissionsSetApproveAllParams(
                            session.getSessionId(), true, null))
                    .get(10, TimeUnit.SECONDS);
            assertTrue(setResult.success(), "setApproveAll should succeed");

            AssistantMessageEvent response = session
                    .sendAndWait(new MessageOptions().setPrompt("Run 'echo test' and tell me what happens"))
                    .get(60, TimeUnit.SECONDS);
            assertNotNull(response);

            // Handler should not have been called since runtime approves all
            assertEquals(0, handlerCallCount[0],
                    "Permission handler should not be called when setApproveAll is enabled");

            session.close();
        }
    }

    /**
     * Verifies that the SDK correctly waits for a slow permission handler before
     * completing tool execution.
     *
     * @see Snapshot: permissions/should_wait_for_slow_permission_handler
     */
    @Test
    void testShouldWaitForSlowPermissionHandler() throws Exception {
        ctx.configureForTest("permissions", "should_wait_for_slow_permission_handler");

        var handlerEntered = new CompletableFuture<Void>();
        var releaseHandler = new CompletableFuture<Void>();

        try (CopilotClient client = ctx.createClient()) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest((request, invocation) -> {
                        handlerEntered.complete(null);
                        return releaseHandler.thenApply(
                                v -> new PermissionRequestResult().setKind(PermissionRequestResultKind.APPROVED));
                    })).get();

            // Capture the sendAndWait future before awaiting it so we can interact with the
            // handler
            CompletableFuture<AssistantMessageEvent> responseFuture = session
                    .sendAndWait(new MessageOptions().setPrompt("Run 'echo slow_handler_test'"));

            // Wait for permission handler to be entered
            handlerEntered.get(30, TimeUnit.SECONDS);

            // Release the handler
            releaseHandler.complete(null);

            // Session should complete successfully
            AssistantMessageEvent message = responseFuture.get(60, TimeUnit.SECONDS);
            assertNotNull(message);

            session.close();
        }
    }
}
